designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,203 @@
+
+* Pros
+  * `kubectl` and `git` follow the same approach for discovering plugins, so there’s prior art.
+
+  * There’s a lot of flexibility in just dropping plugin binaries to PATH variable and enabling the discovery without having to enforce any other constraints on the placements of the plugins.
+
+* Cons
+  * Enumerating the list of all available plugins might be a bit tough compared to having a single folder with the list of available plugins and having to enumerate those.
+
+  * These plugin binaries cannot be run in a standalone manner outside of Kubebuilder, so may not be very ideal to add them to the PATH var.
+
+#### What Plugin system should we use
+
+Based on the evaluation of plugin libraries such as the [built-in go-plugin library](https://golang.org/pkg/plugin/) and [Hashicorp’s plugin library](https://github.com/hashicorp/go-plugin), it is more suitable to write our own custom plugin library as the built-in plugin library seems to be more suitable for in-tree plugins rather than out-of-tree plugins and it doesn’t offer cross language support, thereby making it a non-starter. Hashicorp’s go plugin system is more suitable than the built-in go-plugin library as it enables cross language/platform support. However, it is more suited for long running plugins as opposed to short lived plugins and the usage of protobuf could be overkill as we will not be handling 10s of 1000s of deserializations.
+
+For the stated reasons, I propose we use our own plugin system that passes JSON blobs back and forth across stdin/stdout and make stdin/stdout the only default option for now as it’s the most language agnostic way and it makes it easy to work with most languages.
+
+In the future, if a need arises (for e.g. users are hitting performance issues), we can then explore the possibility of using the Hashicorp’s go plugin library. From a design standpoint, to leave it architecturally open, I propose using a “type” field in the PROJECT file to potentially allow other plugin libraries in the future and make this a seperate field in the PROJECT file per plugin; and this field determines how the injector will be passed for a given plugin. However, for the sake of simplicity in initial design and not to introduce any breaking changes as Project version 3 would suffice for our needs, this option is out of scope in this proposal.
+
+#### Project configuration
+
+Currently, the project configuration has two fields to store plugin specific information.
+
+* `Layout` field (of type string) is used for plugin chain resolution on initialized projects. This will be the default if no plugins are specified for a subcommand.
```

Comment:
```suggestion
* `Layout` field (of type []string) is used for plugin chain resolution on initialized projects. This will be the default if no plugins are specified for a subcommand.
```